File: src/rotation_matrix.py
import numpy as np
from math import cos, sin

def _rot_mat_x(angle):
    return np.mat([[1, 0, 0],
                    [0, cos(angle), -sin(angle)],
                    [0, sin(angle), cos(angle)]])

def _rot_mat_y(angle):
    return np.mat([[cos(angle), 0, sin(angle)],
                    [0, 1, 0],
                    [-sin(angle), 0, cos(angle)]])

def _rot_mat_z(angle):
    return np.mat([[cos(angle), -sin(angle), 0],
                    [sin(angle), cos(angle), 0],
                    [0, 0, 1]])

def rotation_matrix(angle):
    rot_x = _rot_mat_x(angle[0])
    rot_y = _rot_mat_y(angle[1])
    rot_z = _rot_mat_z(angle[2])

    rot = np.matmul(rot_x, rot_y)
    rot = np.matmul(rot, rot_z)

    return rot

File: src/lib/rotation_matrix.cpp
#include "rotation_matrix.hpp"

#include <cmath>

using namespace cv;
using namespace std;

static Matx33f _rot_mat_x(float angle){
    Matx33f m({1, 0, 0,
            0, cosf(angle), -sinf(angle),
            0, sinf(angle), cosf(angle)});
    return m;
}

static Matx33f _rot_mat_y(float angle) {
    Matx33f m({cosf(angle), 0, sinf(angle),
            0, 1, 0,
            -sinf(angle), 0, cosf(angle)});
    return m;
}

static Matx33f _rot_mat_z(float angle) {
    Matx33f m({cosf(angle), -sinf(angle), 0,
            sinf(angle), cosf(angle), 0,
            0, 0, 1});
    return m;
}

void rotation_matrix(const vector<float> &angle, Matx33f &rotation_matrix)
{
    Matx33f rot_x = _rot_mat_x(angle[0]);
    Matx33f rot_y = _rot_mat_y(angle[1]);
    Matx33f rot_z = _rot_mat_z(angle[2]);


    rotation_matrix = rot_x.mul(rot_y).mul(rot_z);
}
